### PR TITLE
Restore updatemasternode and combine setforcedrewardaddress / removeforcedaddress

### PR DESCRIPTION
**Pull request by Bushstar**
This PR also changes on the GW fork the stake modifier calculation to use the MN owner address instead of the operator address.

### CODE REVIEW

**Comment by Stonygan** [comment]
Maybe you can check, why its only allowed to use a Legacy or address from the LightWallet(BECH32?) for a rewardaddress? Normal addresses from the APP or CAKE are not allowed? @Bushstar

**Review by bvbfan — APPROVED** [review]

**Comment by mjollnir14** [comment]
Very glad to see this feature back. Awesome job guys.

File: src/masternodes/mn_checks.cpp [hunk 1]
```diff
@@ -922,28 +922,32 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
...
-                return Res::Err("Masternode with that operator address already exists");
-            }
-            mnview.UpdateMasternode(obj.mnId, *node, obj.operatorType, obj.operatorAddress, height);
-        }
-
-        if (obj.secondType == static_cast<uint8_t>(UpdateMasternodeType::SetRewardAddress)) {
-            if (obj.rewardType != 1 && obj.rewardType != 4) {
-                return Res::Err("Reward address must be P2PKH or P2WPKH type");
+        for (const auto& item : obj.updates) {
+            LogPrintf("TYPE: %d\n", item.first);
```
Comment (bvbfan) [inline src/masternodes/mn_checks.cpp:930]: We can remove the log.

**Review by bvbfan — APPROVED** [review]

**Review by surangap — APPROVED** [review]

File: src/masternodes/mn_checks.cpp (hunk 1 above)
Comment (Bushstar) [reply src/masternodes/mn_checks.cpp:930]: Oops... Was trying to figure out why the updated Python test was not working with incorrect type set. Did you spot the empty else statement yet?

**Comment by Bushstar** [comment]
@Stonygan I expect that the initial idea was to simplify ownership. P2PKH and P2WPKH are the Legacy and Bech32 address types, these are simple addresses derived from a private key. The other address type you mention is P2SH-P2WPKH which wraps P2WPKH in P2SH, to accept this we have to accept all P2SH addresses and we do not know how those addresses were created, they could be multi-sig or a different type of script entirely. Better to keep things simple than let people get into trouble.

If you have access to the DeFi CLI console you can generate a legacy or bech32 address with the following commands.

```
getnewaddress "" "legacy"
getnewaddress "" "bech32"
```

**Comment by chanakasameera** [comment]
`MasternodeInfo.banTx` Attribute becomes undefined by this PR, is it acceptable ?
Related jellyfish test : packages/jellyfish-api-core/__tests__/category/masternode/createMasternode.test.ts

**Comment by bvbfan** [comment]
Yes, `banTx` isn't used anywhere, now it's waste of space.

File: src/masternodes/rpc_masternodes.cpp [hunk 2]
```diff
@@ -8,7 +8,7 @@ UniValue mnToJSON(CImmutableCSView& view, uint256 const & nodeId, CMasternode co
     UniValue ret(UniValue::VOBJ);
     auto currentHeight = view.GetLastHeight();
     if (!verbose) {
-        ret.pushKV(nodeId.GetHex(), CMasternode::GetHumanReadableState(node.GetState(currentHeight)));
+        ret.pushKV(nodeId.GetHex(), CMasternode::GetHumanReadableState(node.GetState(currentHeight, *pcustomcsview)));
```
Comment (bvbfan) [inline src/masternodes/rpc_masternodes.cpp:11]: Use view

File: src/masternodes/rpc_masternodes.cpp [hunk 3]
```diff
@@ -30,8 +30,8 @@ UniValue mnToJSON(CImmutableCSView& view, uint256 const & nodeId, CMasternode co
         obj.pushKV("creationHeight", node.creationHeight);
         obj.pushKV("resignHeight", node.resignHeight);
         obj.pushKV("resignTx", node.resignTx.GetHex());
-        obj.pushKV("banTx", node.banTx.GetHex());
-        obj.pushKV("state", CMasternode::GetHumanReadableState(node.GetState(currentHeight)));
+        obj.pushKV("collateralTx", node.collateralTx.GetHex());
+        obj.pushKV("state", CMasternode::GetHumanReadableState(node.GetState(currentHeight, *pcustomcsview)));
```
Comment (bvbfan) [inline src/masternodes/rpc_masternodes.cpp:34]: `view`

File: src/masternodes/rpc_masternodes.cpp [hunk 4]
```diff
@@ -48,7 +48,7 @@ UniValue mnToJSON(CImmutableCSView& view, uint256 const & nodeId, CMasternode co
         uint16_t timelock = view.GetTimelock(nodeId, node, currentHeight);
 
         // Only get targetMultiplier for active masternodes
-        if (node.IsActive(currentHeight)) {
+        if (node.IsActive(currentHeight, *pcustomcsview)) {
```
Comment (bvbfan) [inline src/masternodes/rpc_masternodes.cpp:51]: `view`

**Review by bvbfan — CHANGES_REQUESTED** [review]

**Review by bvbfan — APPROVED** [review]

**Comment by mjollnir14** [comment]
Will this be included in Fort Canning Great World?

**Pull request merged**